sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java

```diff
@@ -89,37 +92,47 @@ public void testGetReadWithKeyPattern() throws Exception {
       pipeline.run();
   }
 
-  @Test
-  public void testGetReadWithNotMatchingKeyPattern() throws Exception {
-    try (EmbeddedRedis embeddedRedis = new EmbeddedRedis()) {
-      insertKV(embeddedRedis);
+  @Test public void testGetReadWithNotMatchingKeyPattern() throws Exception {
+    PCollection<KV<String, String>> read = pipeline.apply(
+        RedisIO.read().withConnectionConfiguration(createConnection(embeddedRedis))
+            .withKeyPattern("foobar"));
 
-      PCollection<KV<String, String>> read = pipeline.apply(
-          RedisIO.read()
-              .withConnectionConfiguration(createConnection(embeddedRedis))
-              .withKeyPattern("foobar"));
+    PAssert.thatSingleton(read.apply("Count", Count.<KV<String, String>>globally())).isEqualTo(0L);
 
-      PAssert.thatSingleton(read.apply("Count", Count.<KV<String, String>>globally()))
-          .isEqualTo(0L);
+    pipeline.run();
+  }
 
-      pipeline.run();
-    }
+  @Test public void testGetReadAll() throws Exception {
+    PCollection<KV<String, String>> readAll = pipeline.apply(Create.of("Key 0", "Key 1"))
+        .apply(RedisIO.readAll().withConnectionConfiguration(createConnection(embeddedRedis)));
+
+    PAssert.thatSingleton(readAll.apply("Count", Count.<KV<String, String>>globally()))
+        .isEqualTo(2L);
+
+    pipeline.run();
   }
 
   @Test
-  public void testGetReadAll() throws Exception {
-    try (EmbeddedRedis embeddedRedis = new EmbeddedRedis()) {
-      insertKV(embeddedRedis);
+  public void testWrite() throws Exception {
+    ArrayList<KV<String, String>> data = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      KV<String, String> kv = KV.of("wkey " + i, "wvalue " + i);
+      data.add(kv);
+    }
+    PCollection<KV<String, String>> input =
+        pipeline.apply(Create.of(data));
 
-      PCollection<KV<String, String>> readAll =
-          pipeline.apply(Create.of("Key 0", "Key 1"))
-          .apply(RedisIO.readAll()
-              .withConnectionConfiguration(createConnection(embeddedRedis)));
+    input.apply(RedisIO.write().withConnectionConfiguration(createConnection(embeddedRedis)));
 
-      PAssert.thatSingleton(readAll.apply("Count", Count.<KV<String, String>>globally()))
-          .isEqualTo(2L);
+    pipeline.run();
 
-      pipeline.run();
+    Jedis jedis = new Jedis("localhost", embeddedRedis.getPort());
```

Comment:
How about doing the verification using a RedisIO.read() pipeline?

Actually: now that you have both read() and write(), you could probably refactor this entire suite into one tiny test:
- Use RedisIO.write() to write data
- Apply: RedisIO.read() to read everything and PAssert on that; RedisIO.read() to read a prefix and PAssert on that; RedisIO.read() to read something that doesn't match any keys and PAssert on that.

https://beam.apache.org/documentation/io/testing/ see "Use a write then read style for your tests." - following that guidance, you can further improve this code by using CountingInput and TestRow. This guidance is under "integration tests" but I think it applies quite well to unit tests too.

Reply:
Yeah, good idea. As it was originally two PRs, I did that way. I will refactore as suggested.